src/Data/Arena.hs

```diff
@@ -0,0 +1,74 @@
+{-# LANGUAGE CPP #-}
+module Data.Arena (
+    ArenaManager,
+    newArenaManager,
+    Arena,
+    Size,
+    Offset,
+    Alignment,
+    withArena,
+
+    allocateFromArena,
+    -- * Test helpers
+    withUnmanagedArena,
+) where
+
+import           Control.DeepSeq (NFData (..))
+import           Control.Monad.Primitive
+import           Data.Primitive.ByteArray
+import           Data.Primitive.MutVar
+
+#ifdef NO_IGNORE_ASSERTS
+import           Control.Monad (forM_)
+import           Data.Word (Word8)
+#endif
+
+data ArenaManager s = ArenaManager
+
+newArenaManager :: PrimMonad m => m (ArenaManager (PrimState m))
+newArenaManager = do
+    _toUseTheConstraint <- newMutVar 'x'
+    return ArenaManager
+
+-- | For use in bencmark environments
+instance NFData (ArenaManager s) where
+    rnf _ = ()
+
+-- TODO: this is debug implementation
+-- we retain all the allocated arrays,
+-- so we can scramble them at the end.
+data Arena s = Arena (MutVar s [MutableByteArray s])
+
+type Size      = Int
+type Offset    = Int
+type Alignment = Int
+
+withArena :: PrimMonad m => ArenaManager (PrimState m) -> (Arena (PrimState m) -> m a) -> m a
+withArena _ f = do
+    mvar <- newMutVar []
+    result <- f (Arena mvar)
+
+    -- scramble the allocated bytearrays,
+    -- they shouldn't be in use anymore!
+#ifdef NO_IGNORE_ASSERTS
+    mbas <- readMutVar mvar
+    forM_ mbas $ \mba -> do
+        size <- getSizeofMutableByteArray mba
+        setByteArray mba 0 size (0x77 :: Word8)
+#endif
+
+    return result
+
+-- | Create unmanaged arena
+--
+-- Never use this in non-tests code.
+withUnmanagedArena :: PrimMonad m => (Arena (PrimState m) -> m a) -> m a
+withUnmanagedArena k = do
+    mgr <- newArenaManager
+    withArena mgr k
+
+allocateFromArena :: PrimMonad m => Arena (PrimState m)-> Size -> Alignment -> m (Offset, MutableByteArray (PrimState m))
+allocateFromArena (Arena mvar) !size !alignment = do
```

Comment:
So this is quite general, supporting arbitrary size and alignment. (I know that's what an arena allocator can do in general.)

Our use case is simpler, we should just need allocations that are all of the same size (4k) and _not_ need to be able to allocate contiguous spans of pages.

Yes, currently the `indexSearches` currently does a single IOOpRead for a span of pages into a single contigious buffer, but I think we should aim to change that to issue multiple reads for each individual page into separate buffers.

That way the underlying allocator only has to deal with a single size of allocation, and so we can use a really simple free-list style.

Reply:
> That way the underlying allocator only has to deal with a single size of allocation, and so we can use a really simple free-list style.

There is no issue with supporting arbitrary allocation/alignment size though.

If the allocator reserves e.g. 1M, and then just keeps the pointer to the "rest/free" space, it's not difficult.

Recall, this allocator doesn't support individuals `free`s. Everything is released at once. And that what makes it simple (and hopefully fast).

Reply:
> Should we change that first then?

With your flexible arena allocator, it would be possible to support the existing `indexSearches`, so we could introduce the changes in either order.